Add GenCode Gene track and show exon number in IGV Show exon number in IGV

northwestwitch opened the pull request:
Fix #2841.
**How to test**:
1. Install this branch on stage and compare gene tracks from master and this branch
2. IGV genes track from this branch should show exon number

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

northwestwitch commented on scout/server/blueprints/alignviewers/templates/alignviewers/utils.html:
```diff
@@ -1,5 +1,5 @@
   <link rel="shortcut icon" href="//igv.org/web/img/favicon.ico">
   <!-- IGV JS-->
-  <script src="https://cdn.jsdelivr.net/npm/igv@2.8.5/dist/igv.min.js"></script>
+  <script src="http://igv.org/web/snapshot/dist/igv.js"></script>
```
This has to be replaced with version 2.10 when it is released

northwestwitch commented:
Before:
![image](https://user-images.githubusercontent.com/28093618/132311042-6c81525f-8e03-45d1-9a41-66b1a57a40a0.png)


After:
![image](https://user-images.githubusercontent.com/28093618/132310798-5cbb3d14-f065-4f25-8a06-00954cb5ba63.png)

northwestwitch commented:
igv.js version 2.10 has been release and this PR is ready for review!

dnil commented:
Do you know anything about differences in the gene tracks?

northwestwitch commented:
> Do you know anything about differences in the gene tracks?

You mean differences introduced by igv with this release or by me, by switching to the gencode tracks?

northwestwitch commented:
Looking at the [release description](https://github.com/igvteam/igv.js/releases/tag/v2.10.0) 2.10 should also fix the problem of the loci slightly off that is probably the error we see in current Scout (igv 2.8.4)

dnil commented:
> Do you know anything about differences in the gene tracks?
> You mean differences introduced by igv with this release or by me, by switching to the gencode tracks?

I was thinking of the difference between the RefSeq and GenCode tracks. It's a completely separate annotation, isn't it, though rooted in ENSMBL and RefSeq?

dnil commented:
One can apparently publish entire papers on this, so perhaps a manuscript version? 😆 
https://pubmed.ncbi.nlm.nih.gov/26110515/

northwestwitch commented:
> One can apparently publish entire papers on this, so perhaps a manuscript version? 😆
> https://pubmed.ncbi.nlm.nih.gov/26110515/

Glad that you found your answer 😆 
Looks like the GenCode one contains also transcript info

northwestwitch commented:
I was thinking that perhaps it would be better to use a GenCode track newer than 24 (24 is hosted on the Broad Institute Amazon bucket). In that case we should download it from UCSC and host it on our figshare

dnil commented:
Well, RefSeq as such certainly does too. I don't know: just wanted to check if you had any particular reason for it. It would be optimal to reuse the gene and transcript definitions from VEP that were used for annotation, but the "authoritative" version from refSeq has also had its benefits. Gencode to me feels more like a researchers delight than the standard clinical set, but it may have evolved lately.

northwestwitch commented:
> ell, RefSeq as such certainly does too. I don't know: just wanted to check if you had any particular reason for it. It would be optimal to reuse the gene and transcript definitions from VEP that were used for annotation, but the "authoritative" version from refSeq has also had its benefits. Gencode to me feels more like a researchers delight than the standard clinical set, but it may have evolved lately.

The reason is that igv.js worked with the gencode one. Didn't work with out track, but I could try again I guess.

dnil commented:
The "UCSC Genes" track from IGV is perhaps most similar if the standard RefSeq one is not compatible with the show exons feature?

northwestwitch commented:
Ok, I've just created and loaded 4 tracks on our ( my actually ) [figshare space](https://figshare.com/articles/dataset/NCBI_RefSeq_Human_Genes_tracks/16601285):

- Refseq genes GRCh37
- Refseq genes GRCh38
- GenCode GRCh37 (v.38 comprehensive lifted over)
- GenCode GRCh38 (v.38 comprehensive)

So we can test. The GenCode tracks are much bigger than the RefSeq ones, and it makes sense when you check that paper that you mentioned in one of the comments above @dnil. It says:
`GENCODE has a higher proportion of manually annotated gene models than RefSeq and includes more novel splicing features. Given our modern understanding of 'pervasive transcription', one could question to what extent this excess transcription is truly functional, as opposed to potential 'noise'. We have demonstrated that the novel exons and introns annotated by GENCODE and RefSeq share characteristics of transcription with those features already annotated in both sets, suggesting that transcriptional noise is unlikely to be the major explanation for the existence of such transcripts, or at least no more so than for transcripts already independently added to both genesets. The additional coverage and diversity of GENCODE Comprehensive transcripts leads to the identification of many more genic variants than RefSeq, however, transcriptional complexity can also make variant interpretation more difficult (see Additional file 14: Figure S9). `

northwestwitch commented on scout/constants/igv_tracks.py:
```diff
@@ -1,32 +1,42 @@
-HG19GENES_INDEX_URL = (
-    "https://s3.amazonaws.com/igv.broadinstitute.org/annotations/hg19/genes/refGene.hg19.bed.gz.tbi"
-)
+HG19GENES_URL = " https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/30726131/NCBI_Refseq_genes_GRCh37_sorted.bed.gz"
```
I've replace the links to the old RefSeq genes with my figshare links, same also below with the new GenCode tracks. The problem is that it says that access is denied:

![image](https://user-images.githubusercontent.com/28093618/132846069-36ac0ea4-0df2-471c-80e2-774828a2b2b4.png)

But it's weird, since the access to these tracks should be public, here: https://figshare.com/articles/dataset/NCBI_RefSeq_Human_Genes_tracks/16601285?file=30726191

If you are wondering how I can see that these are the right links to the files, try to use wget with the download links provided on the figshare pages

northwestwitch commented:
I grew tired of not being able to use direct links in figshare so I moved all the gene tracks to my IBM cloud storage and it worked immediately.

I've changed the code to support both RefSeq genes (it works in showing exon number now) and GenCode genes. The user can choose which track to use:

![image](https://user-images.githubusercontent.com/28093618/132855325-8e283d19-73ec-433b-ae69-294f9730d974.png)


![image](https://user-images.githubusercontent.com/28093618/132855285-77d14c64-1a90-4935-8776-3405ad71eeca.png)

northwestwitch commented:
@dnil I've requested your review because you commented before, but it's not urgent!

northwestwitch commented:
Closing this, as I am not sure of the PR myself and it's only collecting conflicts for now!

dnil commented:
Sorry, I haven't gotten around to testing it, and am not in the loop on current developments in gene models on the human genome. Perhaps let's set up a small project on this, and see if we can find someone who actually knows the status of the current annotation projects, or some more current papers to look at?

northwestwitch commented:
> Sorry, I haven't gotten around to testing it, and am not in the loop on current developments in gene models on the human genome. Perhaps let's set up a small project on this, and see if we can find someone who actually knows the status of the current annotation projects, or some more current papers to look at?

Sounds good!
